File: android/app/src/main/java/com/gameshelf/ui/wishlist/WishlistSortOrder.kt
package com.gameshelf.ui.wishlist

import androidx.annotation.StringRes
import com.gameshelf.R
import com.gameshelf.data.itad.GameDeal
import com.gameshelf.data.itad.GamePrices
import com.gameshelf.domain.Game
import com.gameshelf.domain.NameOrder
import com.gameshelf.ui.library.SortIcon

/** Como ordenar la lista de deseos. */
enum class WishlistSortOrder(
  val id: String,
  @StringRes val displayNameRes: Int,
  val icon: SortIcon,
) {
  /** Lo ultimo que quisiste, primero. */
  DESEADO_HACE_POCO("deseadoHacePoco", R.string.wishlist_sort_recent, SortIcon.CALENDAR),

  /** La rebaja mas grande, primero. */
  MAYOR_DESCUENTO("mayorDescuento", R.string.wishlist_sort_discount, SortIcon.TAG),

  /** El mas barato, primero. */
  PRECIO_MAS_BAJO("precioMasBajo", R.string.wishlist_sort_price, SortIcon.ARROW_DOWN);

  /**
   * Si el orden depende de tener los precios cargados.
   *
   * Sirve para avisar en vez de mostrar una lista que parece desordenada
   * mientras los precios todavia estan en camino.
   */
  val needsPrices: Boolean get() = this != DESEADO_HACE_POCO

  /**
   * Ordena los juegos.
   *
   * Los juegos sin precio van siempre al final, sin importar el orden: uno del
   * que no se sabe el precio no es "el mas barato", es uno sin datos. Entre
   * ellos se ordenan por nombre para que la lista no baile en cada carga.
   */
  fun sort(juegos: List<Game>, precios: Map<Int, GamePrices>): List<Game> = when (this) {
    DESEADO_HACE_POCO -> porFechaDeseado(juegos)

    MAYOR_DESCUENTO -> ordenar(juegos, precios) { izq, der ->
      der.second.discountPercent.compareTo(izq.second.discountPercent)
    }

    PRECIO_MAS_BAJO -> ordenar(juegos, precios) { izq, der ->
      izq.second.price.amount.compareTo(der.second.price.amount)
    }
  }

  /**
   * Separa los que tienen precio de los que no, ordena los primeros con el
   * criterio dado y deja los segundos al final.
   */
  private fun ordenar(
    juegos: List<Game>,
    precios: Map<Int, GamePrices>,
    criterio: Comparator<Pair<Game, GameDeal>>,
  ): List<Game> {
    val conPrecio = mutableListOf<Pair<Game, GameDeal>>()
    val sinPrecio = mutableListOf<Game>()

    juegos.forEach { juego ->
      val mejor = juego.steamAppID?.let { precios[it]?.best }
      if (mejor != null) conPrecio += juego to mejor else sinPrecio += juego
    }

    return conPrecio.sortedWith(criterio).map { it.first } +
      sinPrecio.sortedWith(NameOrder.byName())
  }

  /** Los mas recientes primero; los que no traen fecha, al final por nombre. */
  private fun porFechaDeseado(juegos: List<Game>): List<Game> {
    val (conFecha, sinFecha) = juegos.partition { it.wishlistedAt != null }
    return conFecha.sortedByDescending { it.wishlistedAt } +
      sinFecha.sortedWith(NameOrder.byName())
  }

  companion object {
    val DEFAULT = DESEADO_HACE_POCO

    fun fromId(id: String?): WishlistSortOrder = entries.firstOrNull { it.id == id } ?: DEFAULT
  }
}

File: GameShelf/ViewModels/WishlistSortOrder.swift
//
//  WishlistSortOrder.swift
//  GameShelf
//

import Foundation

/// Como ordenar la lista de deseos.
enum WishlistSortOrder: String, CaseIterable, Sendable {
  /// Lo ultimo que quisiste, primero.
  case deseadoHacePoco
  /// La rebaja mas grande, primero.
  case mayorDescuento
  /// El mas barato, primero.
  case precioMasBajo

  static let `default` = WishlistSortOrder.deseadoHacePoco

  var displayName: String {
    switch self {
    case .deseadoHacePoco: String(localized: "Deseado hace poco", comment: "Orden de la lista de deseos")
    case .mayorDescuento: String(localized: "Mayor descuento", comment: "Orden de la lista de deseos")
    case .precioMasBajo: String(localized: "Precio mas bajo", comment: "Orden de la lista de deseos")
    }
  }

  var symbolName: String {
    switch self {
    case .deseadoHacePoco: "calendar.badge.clock"
    case .mayorDescuento: "tag"
    case .precioMasBajo: "arrow.down.circle"
    }
  }

  /// Si el orden depende de tener los precios cargados.
  ///
  /// Sirve para avisar en vez de mostrar una lista que parece desordenada
  /// mientras los precios todavia estan en camino.
  var needsPrices: Bool {
    self != .deseadoHacePoco
  }

  /// Ordena los juegos.
  ///
  /// - Parameters:
  ///   - juegos: Los juegos a ordenar.
  ///   - precios: Precios ya cargados, indexados por appid de Steam.
  ///
  /// Los juegos sin precio van siempre al final, sin importar el orden: uno del
  /// que no se sabe el precio no es "el mas barato", es uno sin datos. Entre
  /// ellos se ordenan por nombre para que la lista no baile en cada carga.
  func sort(_ juegos: [Game], precios: [Int: GamePrices]) -> [Game] {
    switch self {
    case .deseadoHacePoco:
      return juegos.sorted(by: porFechaDeseado)

    case .mayorDescuento:
      return ordenar(juegos, precios: precios) { izq, der in
        izq.mejor.discountPercent > der.mejor.discountPercent
      }

    case .precioMasBajo:
      return ordenar(juegos, precios: precios) { izq, der in
        izq.mejor.price.amount < der.mejor.price.amount
      }
    }
  }

  /// Separa los que tienen precio de los que no, ordena los primeros con el
  /// criterio dado y deja los segundos al final.
  private func ordenar(
    _ juegos: [Game],
    precios: [Int: GamePrices],
    por criterio: ((juego: Game, mejor: GameDeal), (juego: Game, mejor: GameDeal)) -> Bool
  ) -> [Game] {
    var conPrecio: [(juego: Game, mejor: GameDeal)] = []
    var sinPrecio: [Game] = []

    for juego in juegos {
      if let appID = juego.steamAppID, let mejor = precios[appID]?.best {
        conPrecio.append((juego, mejor))
      } else {
        sinPrecio.append(juego)
      }
    }

    return conPrecio.sorted(by: criterio).map(\.juego)
      + sinPrecio.sorted { $0.name.localizedStandardCompare($1.name) == .orderedAscending }
  }

  /// Los mas recientes primero; los que no traen fecha, al final por nombre.
  private func porFechaDeseado(_ izq: Game, _ der: Game) -> Bool {
    switch (izq.wishlistedAt, der.wishlistedAt) {
    case let (fechaIzq?, fechaDer?): fechaIzq > fechaDer
    case (nil, _?): false
    case (_?, nil): true
    case (nil, nil): izq.name.localizedStandardCompare(der.name) == .orderedAscending
    }
  }
}

/// Preferencias de la lista de deseos que sobreviven al cierre de la app.
@Observable
@MainActor
final class WishlistPreferences {

  private let defaults: UserDefaults
  private static let sortKey = "wishlist.sortOrder"

  /// Orden elegido. Al cambiarlo se guarda solo.
  var sortOrder: WishlistSortOrder {
    didSet {
      guard sortOrder != oldValue else { return }
      defaults.set(sortOrder.rawValue, forKey: Self.sortKey)
    }
  }

  init(defaults: UserDefaults = .standard) {
    self.defaults = defaults

    let guardado = defaults.string(forKey: Self.sortKey)
    self.sortOrder = guardado.flatMap(WishlistSortOrder.init(rawValue:)) ?? .default
  }
}
